snuba/consumer.py

```diff
@@ -110,3 +134,209 @@ def flush_batch(self, batch: Sequence[ProcessedMessage]):
+    ProcessingStep[Union[None, InsertBatch, ReplacementBatch]]
+):
+    def __init__(
+        self,
+        insert_batch_writer: InsertBatchWriter,
+        replacement_batch_writer: Optional[ReplacementBatchWriter] = None,
+    ) -> None:
+        self.__insert_batch_writer = insert_batch_writer
+        self.__replacement_batch_writer = replacement_batch_writer
+
+        self.__closed = False
+
+    def poll(self) -> None:
+        self.__insert_batch_writer.poll()
+
+        if self.__replacement_batch_writer is not None:
+            self.__replacement_batch_writer.poll()
+
+    def submit(
+        self, message: Message[Union[None, InsertBatch, ReplacementBatch]]
+    ) -> None:
+        assert not self.__closed
+
+        if message.payload is None:
+            return
+
+        if isinstance(message.payload, InsertBatch):
+            self.__insert_batch_writer.submit(cast(Message[InsertBatch], message))
+        elif isinstance(message.payload, ReplacementBatch):
+            assert self.__replacement_batch_writer is not None
+            self.__replacement_batch_writer.submit(
+                cast(Message[ReplacementBatch], message)
+            )
+        else:
+            raise TypeError("unexpected payload type")
+
+    def close(self) -> None:
+        self.__closed = True
+
+        self.__insert_batch_writer.close()
+
+        if self.__replacement_batch_writer is not None:
+            self.__replacement_batch_writer.close()
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+        self.__insert_batch_writer.join(timeout)
+
+        if self.__replacement_batch_writer is not None:
+            if timeout is not None:
+                timeout = max(timeout - (time.time() - start), 0)
+
+            self.__replacement_batch_writer.join(timeout)
+
+
+class StreamingConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    def __init__(
+        self,
+        prefilter: Optional[StreamMessageFilter[KafkaPayload]],
+        processor: MessageProcessor,
+        writer: BatchWriter[WriterTableRow],
+        max_batch_size: int,
+        max_batch_time: int,
+        replacements_producer: Optional[ConfluentKafkaProducer] = None,
+        replacements_topic: Optional[Topic] = None,
+    ) -> None:
+        self.__prefilter = prefilter
+        self.__processor = processor
+        self.__writer = writer
+
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+
+        assert not (replacements_producer is None) ^ (replacements_topic is None)
+        self.__supports_replacements = replacements_producer is not None
+        self.__replacements_producer = replacements_producer
+        self.__replacements_topic = replacements_topic
+
+    def __should_accept(self, message: Message[KafkaPayload]) -> bool:
+        assert self.__prefilter is not None
+        return not self.__prefilter.should_drop(message)
```

Comment:
Probably will add a `should_accept` method or similar to `StreamMessageFilter` in the future that behaves like a normal predicate function to the built-in `filter` function as well as `FilterStep` to avoid having to invert the result here.